File: Part 2 - Regression/Section 5 - Multiple Linear Regression/my_mlr.py
import numpy as np
import pandas as pd
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import statsmodels.formula.api as sm

# Load the data and separate x and y
dataset = pd.read_csv("50_Startups.csv")
x = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values

# One-hot encode the categorical variable (State)
categories = np.unique(x[:, 3])
enc = OneHotEncoder(categories=[categories])
encoded_values = enc.fit_transform(x[:, 3].reshape(-1, 1)).toarray()
x = np.append(encoded_values.astype(int), x[:, :-1].astype(float), axis=1)

# Avoid Dummy Variable Trap
x = x[:, 1:]

# Split dataset into train and test - 20%
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=0)

# Fit Multiple Linear Regression to the set
regressor = LinearRegression()
regressor.fit(x_train, y_train)
y_pred = regressor.predict(X=x_test)

# Compare the results
comparison = np.append(y_test.reshape(-1, 1), y_pred.reshape(-1, 1), axis=1)
print(comparison, "sum of differences: ", np.sum(abs(y_test - y_pred)))

# Add ones to the beginning of the x (acts as coefficient for constant)
x = np.append(arr=np.ones((50, 1)).astype(int), values=x, axis=1)

# Remove insignificant values
X_opt = x[:, [0, 3]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
print(regressor_OLS.summary())

# New predictions with less variables
regressor_2 = LinearRegression()
regressor_2.fit(x_train[:, [0, 3]], y_train)
y_pred_2 = regressor_2.predict(X=x_test[:, [0, 3]])

# Compare results
comparison = np.append(y_test.reshape(-1, 1), y_pred_2.reshape(-1, 1), axis=1)
print(comparison, "sum of differences: ", np.sum(abs(y_test - y_pred_2)))

File: Part 2 - Regression/Section 5 - Multiple Linear Regression/my_mlr.R
setwd("C:/Users/Ireneusz Larysz/Downloads/Python/Machine_Learning/Part 2 - Regression/Section 5 - Multiple Linear Regression")
dataset <- read.csv("50_Startups.csv")

# Category data will is going encoded as 0,1
# According to country column a "One Hot Encoding" will be used
dataset$State <- factor(dataset$State,
                          levels = as.vector(unique(dataset$State)),
                          labels = c(1,2,3))

# Split dataset into training and test set
library("caTools")
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
# training_set = dataset$Purchased[split]
# test_set = dataset$Purchased[!split]
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Fit regressor to the training set
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend
              + State, data = dataset)
# Another way to write the formula:

regressor = lm(formula = Profit ~ ., data = dataset)

# Only R&D spend turne out to be signifacnt (p < 0.05)
regressor_2 = lm(formula = Profit ~ R.D.Spend, data = dataset)
y_pred = predict(regressor_2, newdata = test_set)








